### PR TITLE
Add KubeletPreferredAddressTypes

### PR DESCRIPTION
**Pull request by vjm**
https://github.com/kubernetes/kops/issues/1074

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/1083)

### CODE REVIEW

**Comment by vjm** [comment]
@k8s-ci-robot I signed it!

File: pkg/apis/kops/componentconfig.go [hunk 1]
```diff
@@ -374,20 +374,21 @@ type KubeAPIServerConfig struct {
...
+	AdmissionControl      		[]string `json:"admissionControl,omitempty" flag:"admission-control"`
+	ServiceClusterIPRange 		string   `json:"serviceClusterIPRange,omitempty" flag:"service-cluster-ip-range"`
+	ClientCAFile          		string   `json:"clientCAFile,omitempty" flag:"client-ca-file"`
+	BasicAuthFile         		string   `json:"basicAuthFile,omitempty" flag:"basic-auth-file"`
+	TLSCertFile           		string   `json:"tlsCertFile,omitempty" flag:"tls-cert-file"`
+	TLSPrivateKeyFile     		string   `json:"tlsPrivateKeyFile,omitempty" flag:"tls-private-key-file"`
+	TokenAuthFile         		string   `json:"tokenAuthFile,omitempty" flag:"token-auth-file"`
+	AllowPrivileged       		*bool    `json:"allowPrivileged,omitempty" flag:"allow-privileged"`
+	APIServerCount        		*int     `json:"apiServerCount,omitempty" flag:"apiserver-count"`
+	KubletPreferredAddressTypes	[]string `json:"kubletPreferredAddressTypes,omitempty" flag:"kubelet-preferred-address-types"`
```
Comment (justinsb) [inline pkg/apis/kops/componentconfig.go]: Typo: `Kublet` vs `Kubelet` (in the field but more importantly in the JSON)

File: upup/models/config/components/kube-apiserver/kube-apiserver.options [hunk 2]
```diff
@@ -16,3 +16,7 @@ KubeAPIServer:
   LogLevel: 2
   AllowPrivileged: true
   Image: {{ Image "kube-apiserver" }}
+  KubletPreferredAddressTypes:
```
Comment (justinsb) [inline upup/models/config/components/kube-apiserver/kube-apiserver.options]: Will need to update this one also

**Comment by vjm** [comment]
@chrislovecnm I signed the CLA, all set there.

@justinsb I fixed these typos, but am still struggling to verify locally. Any guidance on how to check that these flags are being passed? When I add them to the YAML options using `kops -v 5 edit cluster` I don't see them being persisted. Is there a better way to test?

**Comment by vjm** [comment]
@justinsb @chrislovecnm any updates on this?

File: pkg/apis/kops/componentconfig.go [hunk 3]
```diff
@@ -374,20 +374,21 @@ type KubeAPIServerConfig struct {
 
 	LogLevel int `json:"logLevel,omitempty" flag:"v"`
```
Comment (kris-nova) [inline pkg/apis/kops/componentconfig.go]: What's with the formatting here? Not that it should matter if it's just formatting - but do we need this in the PR?

Comment (justinsb) [inline pkg/apis/kops/componentconfig.go]: It's gofmt.  The alternative is an extra line break to separate the fields.

File: upup/models/config/components/kube-apiserver/kube-apiserver.options [hunk 4]
```diff
@@ -16,3 +16,7 @@ KubeAPIServer:
   LogLevel: 2
   AllowPrivileged: true
   Image: {{ Image "kube-apiserver" }}
+  KubeletPreferredAddressTypes:
```
Comment (justinsb) [inline upup/models/config/components/kube-apiserver/kube-apiserver.options]: I've realized you can only pass this in 1.5 or later.

I think you need to put this in a `_k8s_1_5` magic directory (like we do for ConfigureCBR0, though the "other way round").  Or - if these are the defaults - simply not specify them unless you are overriding.

Are we changing the defaults here?

Comment (vjm) [inline upup/models/config/components/kube-apiserver/kube-apiserver.options]: followed the same pattern as `AnonymousAuth`

**Comment by chrislovecnm** [comment]
Just to let you know this https://github.com/kubernetes/kops/pull/1183 will probably impact this PR. I appreciate your patience, and help! This PR is a big change that has some huge benefits.

**Comment by chrislovecnm** [comment]
@justinsb recommended some changes :)

https://github.com/kubernetes/kops/issues/1206

**Comment by vjm** [comment]
@chrislovecnm thanks for the response -- anything I need to be doing?

**Comment by chrislovecnm** [comment]
@vjm we good to go?

**Comment by vjm** [comment]
@chrislovecnm now we are! updated and working

File: Makefile [hunk 5]
```diff
@@ -251,7 +251,7 @@ examples:
 
 apimachinery:
 	#go install ./cmd/libs/go2idl/conversion-gen
-	~/k8s/bin/conversion-gen  --input-dirs k8s.io/kops/pkg/apis/kops/v1alpha1 --v=8  --output-file-base=zz_generated.conversion
+	${GOPATH}/bin/conversion-gen --skip-unsafe=true --input-dirs k8s.io/kops/pkg/apis/kops/v1alpha1 --v=8  --output-file-base=zz_generated.conversion
```
Comment (chrislovecnm) [inline Makefile]: Can we back this out?  If needed can we put it in a separate commit at least?

Comment (chrislovecnm) [inline Makefile]: Not a PR a commit

File: pkg/apis/kops/componentconfig.go [hunk 6]
```diff
@@ -380,6 +380,8 @@ type KubeAPIServerConfig struct {
 	RuntimeConfig map[string]string `json:"runtimeConfig,omitempty" flag:"runtime-config"`
 
 	AnonymousAuth *bool `json:"anonymousAuth,omitempty" flag:"anonymous-auth"`
+	
+	KubeletPreferredAddressTypes []string `json:"kubeletPreferredAddressTypes,omitempty" flag:"kubelet-preferred-address-types"`
```
Comment (chrislovecnm) [inline pkg/apis/kops/componentconfig.go]: should we add documentation on wth this thing is?

File: pkg/apis/kops/v1alpha1/componentconfig.go [hunk 7]
```diff
@@ -377,6 +377,8 @@ type KubeAPIServerConfig struct {
 	RuntimeConfig         map[string]string `json:"runtimeConfig,omitempty" flag:"runtime-config"`
 
 	AnonymousAuth *bool `json:"anonymousAuth,omitempty" flag:"anonymous-auth"`
+
+	KubeletPreferredAddressTypes []string `json:"kubeletPreferredAddressTypes,omitempty" flag:"kubelet-preferred-address-types"`
```
Comment (chrislovecnm) [inline pkg/apis/kops/v1alpha1/componentconfig.go]: ` pkg/apis/kops/v1alpha1/componentconfig.go` I think we need ` pkg/apis/kops/v1alpha2/componentconfig.go`

File: pkg/apis/kops/v1alpha1/zz_generated.conversion.go [hunk 8]
```diff
@@ -199,8 +199,7 @@ func autoConvert_v1alpha1_ClusterSpec_To_kops_ClusterSpec(in *ClusterSpec, out *
 	if in.Topology != nil {
 		in, out := &in.Topology, &out.Topology
 		*out = new(kops.TopologySpec)
-		// TODO: Inefficient conversion - can we improve it?
```
Comment (chrislovecnm) [inline pkg/apis/kops/v1alpha1/zz_generated.conversion.go]: Can you file an issue please.

File: pkg/apis/kops/v1alpha2/componentconfig.go [hunk 9]
```diff
@@ -133,6 +133,10 @@ type KubeAPIServerConfig struct {
 	// for KubeAPIServer, concatenated with commas. ex: `--runtime-config=key1=value1,key2=value2`.
 	// Use this to enable alpha resources on kube-apiserver
 	RuntimeConfig map[string]string `json:"runtimeConfig,omitempty" flag:"runtime-config"`
+
+	AnonymousAuth *bool `json:"anonymousAuth,omitempty" flag:"anonymous-auth"`
+
+	KubeletPreferredAddressTypes []string `json:"kubeletPreferredAddressTypes,omitempty" flag:"kubelet-preferred-address-types"`
```
Comment (chrislovecnm) [inline pkg/apis/kops/v1alpha2/componentconfig.go]: So you have it in both alpha1 and alpha2 ... I would say we just need it in alpha2

File: pkg/apis/kops/componentconfig.go (hunk 6 above)
Comment (justinsb) [inline pkg/apis/kops/componentconfig.go]: Probably not worth it for these - they directly map to the flags.  Eventually this will be part of the k8s API, and we'll be able to delete this one.

File: pkg/apis/kops/v1alpha2/componentconfig.go [hunk 10]
```diff
@@ -133,6 +133,10 @@ type KubeAPIServerConfig struct {
 	// for KubeAPIServer, concatenated with commas. ex: `--runtime-config=key1=value1,key2=value2`.
 	// Use this to enable alpha resources on kube-apiserver
 	RuntimeConfig map[string]string `json:"runtimeConfig,omitempty" flag:"runtime-config"`
+
+	AnonymousAuth *bool `json:"anonymousAuth,omitempty" flag:"anonymous-auth"`
```
Comment (justinsb) [inline pkg/apis/kops/v1alpha2/componentconfig.go]: Oops - nice find :-)

File: pkg/apis/kops/v1alpha1/zz_generated.conversion.go (hunk 8 above)
Comment (justinsb) [inline pkg/apis/kops/v1alpha1/zz_generated.conversion.go]: This is autogen

File: pkg/apis/kops/v1alpha1/componentconfig.go (hunk 7 above)
Comment (justinsb) [inline pkg/apis/kops/v1alpha1/componentconfig.go]: I think it's there?

**Comment by justinsb** [comment]
LGTM - thanks :-)
